### ISSUE TITLE
FR: `length(BigInt, x)` to avoid overflow

### ISSUE THREAD

**Issue by fonsp** [JuliaLang/julia#43422]
For some built-in Julia types, calling the `length` function causes integer overflow:

```julia
julia> Iterators.product(1:1e9, 1:1e9) |> length
1000000000000000000

julia> Iterators.product(1:1e10, 1:1e9) |> length
-8446744073709551616
```

This caused a problem for me, where I was checking _"is this object small enough to send over the network?"_, and the second object gives a false postive. It seems like Base Julia has no built-in way to avoid this.

My feature request would be a to standardise a method `length(T::Type, x)`, where the first argument provides the output type. In my case, I would use it like `length(BigInt, x)`, and the default is `length(Int, x)`.

# Usage
With a fallback `length(T::Type, x) = convert(T, length(x))`, this could gradually be added to the ecosystem:

For example, the new method for the `ProductIterator` would look like:
```julia
length(T::Type, pr::Iterators.ProductIterator) = prod(T[length(T, i) for i in pr.iterators])
```

**Comment by DilumAluthge** [comment]
Alternatively, this particular method of `Base.length` use checked integer arithmetic?

**Comment by fonsp** [comment]
Good idea! I implemented this in #43429. The FR remains (it is still not possible to get the length), but my use case is ("is x small enough for y?") covered by doing checked arithmetic, because I can catch for the `OverflowError`.

**Comment by adienes** [comment]
since the original example now gives an `OverflowError` and can be remedied by using `BigInt` ranges, I think this issue can be closed
